src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -411,13 +411,51 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+            bool hasParameterlessPublicConstructor =
+                constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
+
+            if (constructors.Length > 0 && !hasParameterlessPublicConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                // if the only constructor is private, then throw
+                if (constructors.Length == 1 && !constructors[0].IsPublic)
```

Comment:
We won't need this if we filter to just public instance constructors as mentioned above

Reply:
Good point!